### DISCUSSION TITLE
urql/exchange-auth and multiple retries

### DISCUSSION THREAD

**Discussion by caprica** [urql-graphql/urql#2885] [Q&A] [answered]
I am working with a bespoke authentication scheme, not my decision and I can't change the approach, whereby we have a cache of 3 (or 'N') access tokens.

Implementing authExchange with getAuth, addAuthToOperation and didAuthError I can see the following:

1. First token is fetched in getAuth
2. First token is added as a header in addAuthToOperation
3. didAuthError is invoked and the first error is detected
4. Second token is fetched in getAuth
5. Second token is added as a header in addAuthToOperation
6. didAuthError is invoked and the second error is detected

And that's it, it does not seem to invoke getAuth again for a third token.

I suppose I expected that this would keep going until getAuth returned null, but I am returning different tokens each time.

Can I do something here to keep going, or do you only get one go at a retry?

**Comment by JoviDeCroock** [comment, marked as answer]
You can add the [`retryExchange`](https://formidable.com/open-source/urql/docs/advanced/retry-operations/) but generally I would say that this is the expected behaviour as we can only assume that you are getting invalid tokens because of a reason inherent to your endpoint/some bad input/... which is why there won't be another retry.

I guess you could also keep retrying within the `getAuth` function until you get a valid token 😅 highly unusual as you mention

**Comment by caprica** [comment]
Thanks for your answer, what you describe is perfectly reasonable as expected behaviour, I just need to find a way to make the custom authentication requirement I have work.  If I can do something with retryExchange that would be something, I'll have a look.
